internalshared/configutil/normalize.go

```
@@ -4,88 +4,106 @@
+//	Host:Port            host:port                           or [host]:port
+//	URL                  scheme://user@host/path?query#frag  or scheme://user@[host]/path?query#frag
+//	Destination Address  user@host:port                      or user@[host]:port
+//
+// See:
+//
+//	https://rfc-editor.org/rfc/rfc3986.html
+//	https://rfc-editor.org/rfc/rfc5942.html
+//	https://rfc-editor.org/rfc/rfc5952.html
+func NormalizeAddr(addr string) string {
+	if addr == "" {
 		return ""
 	}
 
-	var ip net.IP
-	var port string
-	bracketedIPv6 := false
+	// Host
+	ip := net.ParseIP(addr)
+	if ip != nil {
+		// net.IP.String() is RFC 5942 §4 compliant
+		return ip.String()
+	}
 
-	// Try parsing it as a URL
-	pu, err := url.Parse(address)
-	if err == nil {
-		// We've been given something that appears to be a URL. See if the hostname
-		// is an IP address
-		ip = net.ParseIP(pu.Hostname())
-	} else {
-		// We haven't been given a URL. Try and parse it as an IP address
-		ip = net.ParseIP(address)
-		if ip == nil {
-			// We haven't been given a URL or IP address, try parsing an IP:Port
-			// combination.
-			idx := strings.LastIndex(address, ":")
-			if idx > 0 {
-				// We've perhaps received an IP:Port address
-				addr := address[:idx]
-				port = address[idx+1:]
-				if strings.HasPrefix(addr, "[") && strings.HasSuffix(addr, "]") {
-					addr = strings.TrimPrefix(strings.TrimSuffix(addr, "]"), "[")
-					bracketedIPv6 = true
-				}
-				ip = net.ParseIP(addr)
-			}
+	// [Host]
+	if strings.HasPrefix(addr, "[") && strings.HasSuffix(addr, "]") {
+		if len(addr) < 3 {
+			return addr
 		}
-	}
 
-	// If our IP is nil whatever was passed in does not contain an IP address.
-	if ip == nil {
-		return address
+		// If we've been given a bracketed IP address, return the address without
+		// normalized without brackets.
+		ip := net.ParseIP(addr[1 : len(addr)-1])
+		if ip != nil {
+			return ip.String()
+		}
+
+		// Our input is not a valid schema.
+		return addr
```

So if I'm understanding this correctly, that means something like `[hashicorp.com]` is invalid and just gets returned as is?

---

Yup. We could return an error in a case where the scheme doesn't match, but that would be adding new requirements. I'm open to that eventually but for now the goal was to do the smallest and least intrusive change to adhere to the spec.

---

Makes sense